### ISSUE TITLE
Can not training with ADAM  "perplexity: nan"

### ISSUE THREAD

**Issue by netxiao** [OpenNMT/OpenNMT#302]
when training model with adam, I have got Perplexity nan.

preprocess.ini
```
src_vocab =./dict/src.dt
tgt_vocab = ./dict/tgt.dt
train_src = ./train/train.src
train_tgt = ./train/train.tgt

valid_src = ./valid/v.src
valid_tgt = ./valid/v.tgt

src_seq_length = 45
tgt_seq_length = 45

```

train.ini
```
optim = adam
learning_rate = 0.0002

layers = 2
rnn_size = 512
gpuid = 1
fp16 = true
curriculum = 1
```

> [06/01/17 13:40:25 INFO]  * vocabulary size: source = 120045; target = 90509
> [06/01/17 13:40:25 INFO]  * additional features: source = 0; target = 0
> [06/01/17 13:40:25 INFO]  * maximum sequence length: source = 45; target = 46
> [06/01/17 13:40:25 INFO]  * number of training sentences: 9625
> [06/01/17 13:40:25 INFO]  * number of batches: 173
> [06/01/17 13:40:25 INFO]    - source sequence lengths: equal
> [06/01/17 13:40:25 INFO]    - maximum size: 64
> [06/01/17 13:40:25 INFO]    - average size: 55.64
> [06/01/17 13:40:25 INFO]    - capacity: 100.00%
> [06/01/17 13:40:25 INFO] Building model...
> [06/01/17 13:40:25 INFO]  * Encoder:
> [06/01/17 13:40:31 INFO]    - word embeddings size: 500
> [06/01/17 13:40:31 INFO]    - type: simple
> [06/01/17 13:40:31 INFO]    - structure: cell = LSTM; layers = 2; rnn_size = 512; dropout = 0.3
> [06/01/17 13:40:31 INFO]  * Decoder:
> [06/01/17 13:40:35 INFO]    - word embeddings size: 500
> [06/01/17 13:40:36 INFO]    - attention: global (general)
> [06/01/17 13:40:36 INFO]    - structure: cell = LSTM; layers = 2; rnn_size = 512; dropout = 0.3
> [06/01/17 13:40:36 INFO]  * Bridge: copy
> [06/01/17 13:40:42 INFO] Initializing parameters...
> [06/01/17 13:40:52 INFO]  * number of parameters: 161898965
> [06/01/17 13:40:52 INFO] Preparing memory optimization...
> [06/01/17 13:40:52 INFO]  * sharing 69% of output/gradInput tensors memory between clones
> [06/01/17 13:40:52 INFO] Start training from epoch 1 to 13...
> [06/01/17 13:40:52 INFO] 
> [06/01/17 13:41:25 INFO] Epoch 1 ; Iteration 50/173 ; Optim ADAM LR 0.0002 ; Source tokens/s 603 ; Perplexity nan

**Comment by guillaumekln** [comment]
And without FP16? You may need a larger learning rate in this precision.

**Comment by netxiao** [comment]
it is FP16 problem. if remove `-fp16` option, worked for me,  I have tested learning rate from 0.0002 to 1, still appear `Perplexity nan`

**Comment by guillaumekln** [comment]
Did you try with a smaller `-max_grad_norm`?

**Comment by netxiao** [comment]
yes, I have tried. but it still show "`Perplexity nan`".

**Comment by guillaumekln** [comment]
What is your GPU?

**Comment by netxiao** [comment]
GTX1060

**Comment by guillaumekln** [comment]
It seems we should keep a FP32 version of the parameters for Adam. However, that's a very low priority as FP16 is not very useful right now (the memory gain is interesting though).

You should use SGD for now if you plan to continue using FP16.

**Comment by netxiao** [comment]
many thanks!
